### PR TITLE
Initial implementation of RomFS filesystem and fsp-srv

### PR DESCRIPTION
**Pull request by bunnei**
This is the initial implementation of `fsp-srv`, also known as the Filesystem service. This change includes the bare minimum implementation to read game ROM file systems (RomFS). These are typically dumped as a decrypted ".istorage" file, alongside the various ExeFS NSOs (main, sdk, rtld, etc.). Furthermore, there are a lot of infrastructure/framework changes here to hopefully make the yuzu filesystem implementation flexible and extensible for the future. A rough overview of what's in this change and how it works:
* We've reused the Citra Loader and FileSys frameworks, which abstracts the "RomFS" concept on a per-loader basis. For now, we implement RomFS for the new` DeconstructedRomDirectory` loader by searching for the first .istorage file in the loaded application's directory.
* Within the FileSys framework, we now have the `FileSystemBackend` interface. This serves as a backend implementation base class for the Switch `IFileSystem` interface (which exposes such things as `CreateFile`, `OpenFile`, `DeleteFile`, `CreateDirectory`, etc.).
* Furthermore, we now have the `StorageBackend` interface. This serves as a backend implementation base class for the Switch `IStorage` interface (which exposes such things as `Read`, `Write`, `GetSize`, `Close` etc.). For example, a `FileSystemBackend::OpenFile` returns a `StorageBackend`, which is analogous to the `IFileSystem`/`IStorage` API.
* I've implemented `RomFS_FileSystem`, which is a very simple implementation of `FileSystemBackend` for the `DeconstructedRomDirectory` RomFS, and furthermore `RomFS_Storage` for the `StorageBackend`.
* Within the `fsp-srv` implementation, `OpenDataStorageByCurrentProcess` opens the `RomFS_FileSystem` file system, opens a `RomFS_Storage` interface from that, then passes it as the `StorageBackend` to an `IStorage` interface.
* `IStorage` implements "Read", which simply calls the `Read` on it's `StorageBackend`, as well has handles any IPC-specific stuff.

Much like the Citra implementation, this should all be abstracted fairly nicely to allow us to do such things as have multiple ROM formats (NSO, NCA, XCI, etc.) without changing the frontend implementation, reuse the same `IStorage` interface with different storage backends (RomFS, savedata, SDMC, etc.), etc.

Probably easiest to review commit-by-commit.

### CODE REVIEW

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        LOG_DEBUG(Service, "called, offset=0x%llx, length=0x%llx", offset, length);
+
+        auto descriptor = ctx.BufferDescriptorB()[0];
+        std::vector<u8> output(length);
+
+        ResultVal<size_t> res = backend->Read(offset, length, output.data());
+        if (res.Failed()) {
+            IPC::RequestBuilder rb{ctx, 2};
+            rb.Push(res.Code());
+        }
```
Comment (Rozelette) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: Missing return

File: src/core/hle/service/filesystem/filesystem.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include "common/common_types.h"
+#include "core/file_sys/filesystem.h"
+#include "core/hle/result.h"
+#include "core/hle/service/service.h"
```
Comment (lioncash) [inline src/core/hle/service/filesystem/filesystem.h]: service.h and filesystem can likely be replaced with forward declarations.

File: src/core/hle/service/filesystem/fsp_srv.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2018 yuzu emulator team
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "core/hle/service/service.h"
```
Comment (lioncash) [inline src/core/hle/service/filesystem/fsp_srv.h]: `<memory>` needs to be included.

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+void FSP_SRV::Initalize(Kernel::HLERequestContext& ctx) {
+    IPC::RequestBuilder rb{ctx, 2};
+    rb.Push(RESULT_SUCCESS);
+    LOG_WARNING(Service, "(STUBBED) called");
+}
+
+void FSP_SRV::GetGlobalAccessLogMode(Kernel::HLERequestContext& ctx) {
+    IPC::RequestBuilder rb{ctx, 4};
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: you're only pushing 3 words to the buffer

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 5]
```diff
@@ -32,49 +31,40 @@ class IStorage final : public ServiceFramework<IStorage> {
...
+        const auto& descriptor = ctx.BufferDescriptorB()[0];
 
-        LOG_DEBUG(Service, "called, offset=0x%llx, length=0x%llx", offset, length);
+        LOG_DEBUG(Service_FS, "called, offset=0x%llx, length=0x%llx", offset, length);
 
-        auto descriptor = ctx.BufferDescriptorB()[0];
-        std::vector<u8> output(length);
+        // Error checking
+        ASSERT_MSG(length == descriptor.Size(), "unexpected size difference");
+        if (length & 0x8000000000000000ULL) {
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: imo just make the length and offset signed and check for < 0

**Pull request merged**